setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/**
+ * Copyright © 2013-2017 Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Setup\Module\Di\Code\Reader;
+
+class FileClassScanner
+{
+    /**
+     * The filename of the file to introspect
+     *
+     * @var string
+     */
+
+    private $filename;
+
+    /**
+     * The list of classes found in the file.
+     *
+     * @var bool
+     */
+
+    private $classNames = false;
+
+    /**
+     * @var array
+     */
+
+    private $tokens;
+
+    /**
+     * Constructor for the file class scanner.  Requires the filename
+     *
+     * @param string $filename
+     */
+
+    public function __construct($filename)
+    {
+        $filename = realpath($filename);
+        if (!file_exists($filename) || !\is_file($filename)) {
+            throw new InvalidFileException(
+                sprintf(
+                    'The file "%s" does not exist or is not a file',
+                    $filename
+                )
+            );
+        }
+        $this->filename = $filename;
+    }
+
+    /**
+     * Retrieves the contents of a file.  Mostly here for Mock injection
+     *
+     * @return string
+     */
+
+    public function getFileContents()
+    {
+        return file_get_contents($this->filename);
+    }
+
+    /**
+     * Extracts the fully qualified class name from a file.  It only searches for the first match and stops looking
+     * as soon as it enters the class definition itself.
+     *
+     * Warnings are suppressed for this method due to a micro-optimization that only really shows up when this logic
+     * is called several millions of times, which can happen quite easily with even moderately sized codebases.
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
```

Comment:
> > So I would like a little guidance on what you would prefer: reduce the individual method complexity or notate MD to ignore CC?

> If reducing the method complexity has performance impact, I'd say go for optimized performance in this case and mark the class as ignored.

@schmengler  Performance and Complexity are two different metrics. And in this specific case these metrics do not correlate. Because high Cyclomatic complexity doesn't lead to high performance and vice versa.

In Magento we have a rule, that any new code should not have SuppressWarnings for CouplingBetween objects and CyclomaticComplexity. Because it's from the beginning has a Technical Debt. 

High CyclomaticComplexity often says that some functionality has not been fully covered with automated testing. Because it's too hard to cover all the possible conditional logic and exit points for this method. 

Moreover such code is fragile and really hard to maintain in future.

Some time ago we even actively used [CRAP](https://testing.googleblog.com/2011/02/this-code-is-crap.html) metrics for this.

So, this method should be decomposed on some smaller once